### PR TITLE
Add Matrix type and add gate map mechanism

### LINKED ISSUE #326: Refactor Gate with new Matrix type and add GateMap
To help plugin developers handle Gates in DQCsim it makes some sense to add some new types to the core which deals with matrices used in gates.

The `dqcsim::common::types::Gate` type is used in the gatestream protocol to communicate gates. Their behaviour is defined by their unitary matrix, which currentlly is encoded as
```rust
Vec<Complex64>
```
This proposal introduces a new `Matrix` type which wraps this `Vec<Complex64>` and adds some utility functions to make plugin development easier.

A proposal for this type and it's API:
```rust
#[derive(Hash, PartialEq)]
struct Matrix {
  data: Vec<Complex64>,
  // pre-computed dimension which corresponds to sqrt(data.len())
  dimension: usize,
}

impl Eq for Matrix {}

impl Matrix {
  pub fn new(elements: Vec<Complexf64>) -> Self;
  pub fn approx_equals(&self, other: &Matrix, epsilon: f64) -> bool;

  // returns new matrix + set of control indices (remaining target order is maintained)
  pub fn strip_control(self) -> (Self, HashSet<usize>); // or Vec<bool> ?

  pub fn get(row: usize, column: usize) -> Option<Complex64>;

  pub fn dimension(&self) -> usize;
}

impl Index<(usize, usize)> for Matrix {}
impl Index<usize> for Matrix {}
// Do we want this?
impl IndexMut<usize> for Matrix {}

impl From<Matrix> for Vec<Complexf64> {}
impl TryFrom<Vec<Complexf64>> for Matrix {}

```
The gate type internally uses the new `Matrix` type. Beware this requires modification of the serde behaviour of `Gate`.

The `Gates` enum is moved behind the `plugin` feature gate (#325).

We also need a `MatrixMap` type to allow plugin developers to map incoming matrices to (our or their) gate definitions. A proposal for it's API:

```rust
struct MatrixMap<T, K> {
  detectors: Vec<(K, Box<dyn Fn(Matrix) -> Option<T>>)>,
  map: RefCell<HashMap<(Matrix, ArbData), (K, T)>>,
}

struct MatrixMapBuilder<T, K> {
  map: MatrixMap<T, K>,
}

impl<T, K> MatrixMapBuilder<T, K> {
  pub fn new() -> Self;
  pub fn with_defaults(self) -> Self;
  pub fn with_detector<F: Fn(Matrix) -> Option<T>>(self, key: K, callback: F) -> Self;
  pub fn finish(self) -> GateMap<T, K>;
}

impl<T,K> MatrixMap<T, K> {
  pub fn builder() -> MatrixMapBuilder<T, K>;

  pub fn len(&self) -> usize;
  pub fn detectors(&self) -> &[K];
  pub fn detect(&self, key: &Matrix) -> Option<(K, T)>;
  pub fn detect_with(&self, key: &Matrix, keys: &[K]) -> Option<(K, T)>; // this does not update the map

  fn insert(&mut self, key: Matrix, value: T) -> Option<T>;
}

impl<T, K> Default for MatrixMap<T, K> {
  fn default() -> Self {
    MatrixMap::builder().with_defaults().finish()
  }
}
```

### PR DESCRIPTION
**Pull request by mbrobbel**
Another attempt to close #326. 

This now targets `master` and shows the complete diff.

### CODE REVIEW

File: rust/src/core/common/converter.rs [hunk 1]
```diff
@@ -0,0 +1,1009 @@
...
+
+impl FromArb for () {
+    fn from_arb(_: &mut ArbData) -> Result<Self> {
+        Ok(())
+    }
+}
+
+impl FromArb for u64 {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
+        let mut args = src.get_args_mut().drain(..1);
```
Comment (mbrobbel) [inline rust/src/core/common/converter.rs]: Needs a len check or fails for `ArbData::default()`.

File: rust/src/core/common/converter.rs [hunk 2]
```diff
@@ -0,0 +1,1009 @@
...
+            .ok_or_else(oe_inv_arg("expected 64-bit integer argument in ArbData"))?;
+        Ok(u64::from_le_bytes(i[..].try_into().ok().ok_or_else(
+            oe_inv_arg("expected 64-bit integer argument in ArbData"),
+        )?))
+    }
+}
+
+impl FromArb for f64 {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
+        let mut args = src.get_args_mut().drain(..1);
```
Comment (mbrobbel) [inline rust/src/core/common/converter.rs]: Needs a len check or fails for `ArbData::default()`.

File: rust/src/core/common/converter.rs [hunk 3]
```diff
@@ -0,0 +1,1009 @@
...
+            .nth(0)
+            .ok_or_else(oe_inv_arg("expected double argument in ArbData"))?;
+        Ok(f64::from_le_bytes(f[..].try_into().ok().ok_or_else(
+            oe_inv_arg("expected double argument in ArbData"),
+        )?))
+    }
+}
+
+impl FromArb for (f64, f64, f64) {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
```
Comment (mbrobbel) [inline rust/src/core/common/converter.rs]: General comment about the implementations of `FromArb`, even if an `Err` is returned, the src object may have been modified.

File: rust/src/core/common/converter.rs [hunk 4]
```diff
@@ -0,0 +1,1009 @@
...
+        let a = f64::from_arb(src)?;
+        let b = f64::from_arb(src)?;
+        let c = f64::from_arb(src)?;
+        Ok((a, b, c))
+    }
+}
+
+impl FromArb for Matrix {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
+        let mut args = src.get_args_mut().drain(..1);
```
Comment (mbrobbel) [inline rust/src/core/common/converter.rs]: Needs a len check or fails for `ArbData::default()`.

File: rust/src/core/common/converter.rs [hunk 5]
```diff
@@ -0,0 +1,1009 @@
...
+
+/// A type that can be converted into (part of) an ArbData object.
+///
+/// This is the reverse of `FromArb`. Generally, constructing an arb from a
+/// number of `ToArb` objects in a certain order should make the objects
+/// recoverable using `FromArb` in the reverse order, assuming all involved
+/// types implement both directions.
+pub trait ToArb {
+    /// Convert self to ArbData parameters and add them to the data object.
+    fn to_arb(self, dest: &mut ArbData);
```
Comment (mbrobbel) [inline rust/src/core/common/converter.rs]: Is there a reason why you don't return the `&mut ArbData`?

Comment (mbrobbel) [reply rust/src/core/common/converter.rs]: Also, does it have to take `self`?

File: rust/src/core/common/gates.rs [hunk 6]
```diff
@@ -1,477 +1,736 @@
...
+    RX(f64),
+    /// Arbitrary rotation around Y-axis with specified angle (θ).
+    RY(f64),
+    /// Arbitrary rotation around Z-axis with specified angle (θ).
+    RZ(f64),
+    /// Arbitrary rotation around Z-axis, global phase chosen such that it
+    /// works as a submatrix for controlled phase operations.
+    Phase(f64),
+    /// Same as Phase, but with θ = π/2^k​.
+    PhaseK(u64),
```
Comment (mbrobbel) [inline rust/src/core/common/gates.rs]: Why don't we use `u32` here?

File: rust/src/core/common/converter.rs [hunk 7]
```diff
@@ -0,0 +1,1117 @@
...
+    type Parameters = Matrix;
+
+    fn detect_matrix(
+        &self,
+        matrix: &Matrix,
+        _epsilon: f64,
+        _ignore_global_phase: bool,
+    ) -> Result<Option<Self::Parameters>> {
+        if let Some(expected) = self.num_qubits {
+            if matrix.num_qubits().unwrap() != expected {
```
Comment (mbrobbel) [inline rust/src/core/common/converter.rs]: This unwrap panics if the `Matrix` returns `None` from `num_qubits` (log2).

Comment (mbrobbel) [reply rust/src/core/common/converter.rs]: Maybe better to validate the `num_qubits` field in the contructor of this type.

File: rust/src/core/common/converter.rs [hunk 8]
```diff
@@ -0,0 +1,1117 @@
...
+    }
+
+    fn construct_matrix(&self, params: &Self::Parameters) -> Result<Matrix> {
+        Ok(UnboundGate::R(params.0, params.1, params.2).into())
+    }
+}
+
+/// Matrix converter object for any matrix of a certain size - simply has the
+/// matrix itself as its parameter type.
+pub struct UMatrixConverter {
```
Comment (mbrobbel) [inline rust/src/core/common/converter.rs]: This has no unitary check so maybe it should be named differently?

File: rust/src/core/common/converter.rs (hunk 1 above)
Comment (jvanstraten) [reply rust/src/core/common/converter.rs]: Fixed in 45b9a6ac

File: rust/src/core/common/converter.rs (hunk 2 above)
Comment (jvanstraten) [reply rust/src/core/common/converter.rs]: Fixed in 45b9a6ac

File: rust/src/core/common/converter.rs (hunk 3 above)
Comment (jvanstraten) [reply rust/src/core/common/converter.rs]: Fixed in 45b9a6ac

File: rust/src/core/common/converter.rs (hunk 4 above)
Comment (jvanstraten) [reply rust/src/core/common/converter.rs]: Fixed in 45b9a6ac

File: rust/src/core/common/converter.rs (hunk 7 above)
Comment (jvanstraten) [reply rust/src/core/common/converter.rs]: Added check to make test succeed, didn't feel like changing Matrix. After all, it's a matrix, not necessarily a quantum gate matrix; we currently only use it as such but at least the name should be changed then. OTOH maybe we should've called it SquareMatrix.

File: rust/src/core/common/converter.rs (hunk 5 above)
Comment (jvanstraten) [reply rust/src/core/common/converter.rs]: > Is there a reason why you don't return the `&mut ArbData`?

Not sure why one would do that?

> Also, does it have to take `self`?

Not to satisfy the borrow checker, reasoning is that semantically the value is moved into the `ArbData`.

File: rust/src/core/common/gates.rs (hunk 6 above)
Comment (jvanstraten) [reply rust/src/core/common/gates.rs]: I just used u64 to be consistent in the size with the doubles everywhere else. You could just as well argue that a u16 is enough as doubles have only 11 exponent bits. It's arbitrary.

**Pull request merged**